### PR TITLE
bpf: Split TCP and non-TCP conntrack tables

### PR DESCRIPTION
**Pull request by joestringer**
Based on ~~#5450~~.

We've observed recently that it's common for bursty non-TCP flows to impact the operation of the datapath with respect to how TCP connections are handled, where for instance frequent legitimate bursts of DNS traffic could crowd out the table, preventing new legitimate TCP connections from establishing.

By splitting the TCP flows from non-TCP flows in the connection tracking tables, there should be better isolation guarantees for how such traffic flows could impact each other.

This piggybacks on recent refactoring in #5480, #5488, #5450 to minimize the set of changes and allow the set of CT maps to be generically updated without hardcoding references to specific maps around the codebase.

The metrics that were added in #5296 are extended to also include L4 protocol (`tcp` / `any`) to account for the split in tables, because there isn't currently any co-ordination between the CT garbage collection runs so it's difficult to properly set histograms or gauges based on two individual events occurring.

At the moment, the existing map path is retained at the same place but newer datapath will only use this for TCP connections. This allows upgrade/downgrade to keep the TCP connections around without interrupting them as the map doesn't need to be recreated. Long-lived UDP connections would be disrupted during upgrade.

Related: #5048

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/5501)

### CODE REVIEW

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
test-me-please

EDIT: Green build.

**Comment by joestringer** [comment]
test-me-please

EDIT: Green build.

**Comment by joestringer** [comment]
test-me-please

File: pkg/maps/ctmap/ctmap.go [hunk 1]
```diff
@@ -45,48 +45,27 @@ var (
...
+	MapNameTCP4Global = MapNameTCP4 + "global"
+
+	// Map names for "any" protocols indicate CT for non-TCP protocols.
+	MapNameAny6       = MapNamePrefix + "_any6_"
+	MapNameAny4       = MapNamePrefix + "_any4_"
+	MapNameAny6Global = MapNameAny6 + "global"
+	MapNameAny4Global = MapNameAny4 + "global"
 
 	MapNumEntriesLocal  = 64000
 	MapNumEntriesGlobal = 1000000
```
Comment (tgraf) [inline pkg/maps/ctmap/ctmap.go]: I suggest we do a 600K TCP and 400K UDP split going to get started and start running that to see how it goes. We'll likely have to make it configurable in the end but that change is a bit more invasive.

**Review by tgraf — DISMISSED** [review]
My main concern is around upgrade and downgrade. An upgrade would simply recreate the maps and thus flush all existing state. That may be acceptable. Downgrade to 1.2 would re-create the all-in-one map I suppose. Downgrade to 1.1 would only work to 1.1.4 I assume?

**Comment by joestringer** [comment]
@tgraf I had in mind right now that the TCP table uses the same path as the existing CT map, so that during upgrade/downgrade those connections would be unaffected. Of course, if we change other attributes like table size then that would then cause recreation like you said.

**Comment by joestringer** [comment]
@tgraf and yes, downgrade to 1.1 would need to go to 1.1.4 or to use `CLEAN_CILIUM_STATE`.

File: pkg/maps/ctmap/ctmap.go (hunk 1 above)
Comment (joestringer) [reply pkg/maps/ctmap/ctmap.go]: Seems reasonable. If we do this, we will force CT map recreation on upgrade so might as well name them better on the filesystem too. (The alternative would involve leaving the existing map's details the same and only use it for TCP connections, which should allow upgrade/downgrade to persist the connections without interruption).

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
test-me-please

**Comment by tgraf** [comment]
@joestringer Given that this would have an upgrade impact. Should we add a CLI flag to the agent to set the size and keep it at 1M for TCP but add a new default to the configmap. That way we can document it easily and leave it up to the user whether to opt into the new behavior at the cost of losing all connection state.

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
@tgraf done, new patch on top. Manually tested with minikube.

**Comment by joestringer** [comment]
Latest patch introduced initialisation race, will push fix shortly.

**Comment by joestringer** [comment]
test-me-please

**Comment by tgraf** [comment]
@joestringer Can you describe the upgrade impact of the latest revision of this PR?

**Review by ianvernon — APPROVED** [review]

**Review by tgraf — APPROVED** [review]

**Comment by tgraf** [comment]
> @joestringer Can you describe the upgrade impact of the latest revision of this PR?

Responding to myself:
The TCP table size is unchanged. An upgrade will not cause any disruption of TCP flows. It is possible that existing UDP flows will see packets dropped.

**Comment by joestringer** [comment]
Made some minor verifier tweaks to fit with the new DNS stuff.

**Comment by joestringer** [comment]
test-me-please

**Pull request merged**